### DISCUSSION TITLE
What does Indexing actually do?

### DISCUSSION THREAD

**Discussion by gkrizek** [ordinals/ord#1829] [Q&A]
This might be a dumb question, but I really don't see is described anywhere. What does indexing do? Is it looking for inscriptions in every block and writing the inscriptions to its own database? What's the difference with the `--index-sats` as well?

Bonus question; It seems like it tries to sync the index on every command. Is there a way to make Ord be a long running process and always say be syncing the index?

**Comment by stevel408** [comment]
I'm new to this too, but did get a chance to look through the code, and can perhaps share some thoughts. With regard to your first question, indexing is basically a bunch of tables facilitating various kind of look-ups. Here's a bit of code from `index.rs`:

```
define_table! { HEIGHT_TO_BLOCK_HASH, u64, &BlockHashValue }
define_table! { INSCRIPTION_ID_TO_INSCRIPTION_ENTRY, &InscriptionIdValue, InscriptionEntryValue }
define_table! { INSCRIPTION_ID_TO_SATPOINT, &InscriptionIdValue, &SatPointValue }
define_table! { INSCRIPTION_NUMBER_TO_INSCRIPTION_ID, u64, &InscriptionIdValue }
define_table! { OUTPOINT_TO_SAT_RANGES, &OutPointValue, &[u8] }
define_table! { OUTPOINT_TO_VALUE, &OutPointValue, u64}
define_table! { SATPOINT_TO_INSCRIPTION_ID, &SatPointValue, &InscriptionIdValue }
define_table! { SAT_TO_INSCRIPTION_ID, u64, &InscriptionIdValue }
define_table! { SAT_TO_SATPOINT, u64, &SatPointValue }
define_table! { STATISTIC_TO_COUNT, u64, u64 }
define_table! { WRITE_TRANSACTION_STARTING_BLOCK_COUNT_TO_TIMESTAMP, u64, u128 }
```

You can see pretty clearly from the table names what kind of lookups they help with. With regard to the `--index-sats` option, I was a bit confused first, but then I realized that it is only applicable for the `index` subcommand, controlling whether the `OUTPOINT_TO_SAT_RANGES` table should be created. I assume this might be because this table might be big. You can see this bit of code in in `index.rs`:

```
 if options.index_sats {
          tx.open_table(OUTPOINT_TO_SAT_RANGES)?
            .insert(&OutPoint::null().store(), [].as_slice())?;
        }
```

And later, whenever some operation is attempted, it might check to see if this table exists, which will indicated whether the system "has sat index":

```
  pub(crate) fn has_sat_index(&self) -> Result<bool> {
    match self.begin_read()?.0.open_table(OUTPOINT_TO_SAT_RANGES) {
      Ok(_) => Ok(true),
      Err(redb::Error::TableDoesNotExist(_)) => Ok(false),
      Err(err) => Err(err.into()),
    }
  }
```

Hope this helps...
